python_modules/dagster/dagster/_experimental/staging_io_manager.py

```diff
@@ -0,0 +1,96 @@
+    if not dagster_event:
+        return None
+    if not isinstance(dagster_event.event_specific_data, StepMaterializationData):
+        return None
+
+    return (
+        dagster_event.event_specific_data.materialization
+        if isinstance(dagster_event.event_specific_data.materialization, AssetMaterialization)
+        else None
+    )
+
+
+def get_asset_observation_from_event_log_entry(
+    event_log_entry: EventLogEntry,
+) -> Optional[AssetObservation]:
+    if event_log_entry.dagster_event_type != DagsterEventType.ASSET_OBSERVATION:
+        return None
+    dagster_event = event_log_entry.dagster_event
+    # assert dagster_event
+    if not dagster_event:
+        return None
+    # assert isinstance(dagster_event.event_specific_data, AssetObservationData)
+    if not isinstance(dagster_event.event_specific_data, AssetObservationData):
+        return None
+
+    return (
+        dagster_event.event_specific_data.asset_observation
+        if isinstance(dagster_event.event_specific_data.asset_observation, AssetObservation)
+        else None
+    )
+
+
+def get_text_metadata(
+    metadata_entries: Sequence[Union[MetadataEntry, PartitionMetadataEntry]], label
+) -> Optional[str]:
+    for metadata_entry in metadata_entries:
+        entry_to_search = (
+            metadata_entry if isinstance(metadata_entry, MetadataEntry) else metadata_entry.entry
+        )
+        if entry_to_search.label == label and isinstance(
+            entry_to_search.entry_data, TextMetadataValue
+        ):
+            return entry_to_search.entry_data.value
+
+    return None
+
+
+class DualEnvironmentIOManager(IOManager):
+    def __init__(
+        self, *, prod_io_manager, staging_io_manager, env_key="env", staging_label="staging"
+    ):
+        self.prod_io_manager = prod_io_manager
+        self.staging_io_manager = staging_io_manager
+        self.staging_label = staging_label
+        self.env_key = env_key
+
+    def load_input(self, context: InputContext) -> Any:
+        # if in the staging environment load from staging if available, else fall back to prod
+        asset_key = context.asset_key
+        instance = context.instance
+        assert instance
+        event_log_entry = instance.get_latest_materialization_event(asset_key)
+        if event_log_entry:
+            asset_materialization = get_asset_materialization_from_event_log_entry(event_log_entry)
+            if asset_materialization:
+                env_entry = get_text_metadata(
+                    asset_materialization.metadata_entries, label=self.staging_label
+                )
+                if env_entry:
+                    return self.staging_io_manager.load_input(context)
+
+        return self.prod_io_manager.load_input(context)
+
+    def handle_output(self, context: OutputContext, obj: Any) -> None:
+        self.staging_io_manager.handle_output(context, obj)
+        context.add_output_metadata({self.env_key: self.staging_label})
```

Comment:
`add_output_metadata` currently counterintuitively overwrites prior metadata on the context.

Have had a PR out for a while to fix this that I haven't gotten around to merging: https://github.com/dagster-io/dagster/pull/9518.

Reply:
can you merge or do you want me to take over

Reply:
I'll do it. Just rebased